### PR TITLE
Decorate a class

### PR DESCRIPTION
**Pull request by ahzamir**
### in this pull request I have used Inheritance and Decorator to decorate my classes according to these requirements:

- Interface
  - I created a class Nameable.
  - I implemented a method called correct_name that will raise a NotImplementedError.
- I turned my Person class to Nameable
  - My Person class inherits from Nameable
  - This class has a method correct_name implemented. It simply returns the name attribute.
- I prepared the base Decorator.
  - It inherits from Nameable.
  - In the constructor, I assigned a nameable object from params to an instance variable.
  - I implemented the correct_name method that returns the result of the correct_name method of the @nameable.
- I prepared CapitalizeDecorator and TrimmerDecorator
  - For the CapitalizeDecorator:
    - I created a class that inherits from the base Decorator class.
    - I implemented a method correct_name that capitalizes the output of @nameable.correct_name.
  - For the TrimmerDecorator:
    - I created a class that inherits from the base Decorator class.
    - I implemented a method correct_name that makes sure that the output of @nameable.correct_name has a maximum of 10 characters. If it's longer it should trim the word.

### CODE REVIEW

File: person.rb [hunk 1]
```diff
@@ -1,5 +1,8 @@
-class Person
+require './nameable'
+
+class Person < Nameable
   def initialize(age, name = 'Unknown', parent_permission: true)
+    super()
     @id = Random.rand(1...1000)
     @name = name
     @age = age
```
Comment (BenMukebo) [inline person.rb:8]: - [ ]  Great job 👏🏻 however, your code does not work as expected see the screenshot 👇🏻

03ruby

- Please make sure to add a descriptive README file, your README should have instructions on how to run your application locally

**Review by BenMukebo — CHANGES_REQUESTED** [review]
### Hi @ahzamir,

You did a great job so far 👏🏼 👏🏼
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

### To highlight 💪🏻
- GitHub flow was correctly used.  ✔️
- Descriptive PR ✔️
- No linter errors ✔️
 👍  Good job!


## STATUS: CHANGES REQUESTED  ♻️ 
_Please Check the comments under the review and make all required changes_


- [x] Great job using the Microverse templates to write your README l file 👍🏻  but I think you can customize it to your project in order to make it descriptive and more professional. A professional README file should have a project title, project description, list of technology used in the project, instructions to set the project up locally,  and a screenshot of your project. please refer to this [template](https://github.com/microverseinc/readme-template), you can copy and paste the content but remember to customize it to your project.

- The README file It is the first file a person will see when they encounter your project, so it should be fairly brief but detailed. It will make your project stand out from a bunch of others.


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏 💻  
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: trimmer_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+require './decorator'
+
+class TrimmerDecorator < Decorator
+  def correct_name
+    @nameable.correct_name[0, 10] if @nameable.correct_name > 10
```
Comment (ShahierNashaat) [inline trimmer_decorator.rb]: - [x] Kindly be noticed that you should compare the length of the string not the actual string.

```ruby
@nameable.correct_name[0, 10] if @nameable.correct_name.length > 10
```

**Review by ShahierNashaat — CHANGES_REQUESTED** [review]
Hi @ahzamir, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
:heavy_check_mark: No Linter Error
:heavy_check_mark: Good Pull Request Description

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: trimmer_decorator.rb (hunk 2 above)
Comment (ahzamir) [reply trimmer_decorator.rb]: I solved it thank you very much

**Review by keroloussamy — APPROVED** [review]
Hi @ahzamir, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To highlight 🎯

- No linter errors.✔️
- Good code design and implementation. ✔️
- Add all the requirements. ✔️

clap

## Optional suggestions
_Check the comments under the review._

- [ ] x

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by ahzamir** [comment]
@ShahierNashaat  @keroloussamy  @BenMukebo  Thank you very much for your kind review and feedback. 😊👍

**Pull request merged**